### DISCUSSION TITLE
What is the integer "d" in the ECC.construct() method?

### DISCUSSION THREAD

**Discussion by TechPro424** [Legrandin/pycryptodome#840] [Q&A]
Hi all

In the documentation for the `Crypto.PublicKey.ECC.construct()` method, https://www.pycryptodome.org/src/public_key/ecc#Crypto.PublicKey.ECC.construct
it is specified that the integer `d` must be within the range `[1, order - 1]`

What is the integer `d`?
From here https://www.pycryptodome.org/src/public_key/ecc#Crypto.PublicKey.ECC.EccKey
> d (integer) – A scalar that represents the private component in NIST P curves. It is smaller than the order of the generator point.

What exactly is the "private component"?
Can't find any results for that, even on the web

**Comment by Legrandin** [comment]
> What is the integer d?
> What exactly is the "private component"?

The integer 'd' is what makes an ECC private key private.
It is a very large, hard-to-guess number.
It is called 'd' because that is the same name used in FIPS 186.

The method `Crypto.PublicKey.ECC.generate` will create the right one for you.

Use `Crypto.PublicKey.ECC.generate` only if the private key was generated somewhere else, it cannot be imported via the normal means (using `import_key`) and all you have is the integer value `d`.
